### PR TITLE
WIP: Start RPC API implementation

### LINKED ISSUE #466: Implement Initial RPC Structure.

### PR DESCRIPTION
**Pull request by mslipper**
closes: #466 
<hr>

- [x] Linked to github-issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Updated all relevant documentation (`docs/`)
- [x] Updated all relevant code comments
- [x] Wrote tests
- [ ] Added entries in `PENDING.md` that include links to the relevant issue or PR that most accurately describes the change.
___________________________________
For Admin Use:
- [x] Added appropriate labels to PR (ex. wip, ready-for-review, docs)
- [x] Reviewers Assigned
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

File: cmd/ethermintd/root.go [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+	"github.com/tendermint/ethermint/server"
+	"fmt"
+	"os"
+	"os/signal"
+	"syscall"
+	"context"
+)
+
+var rootCmd = &cobra.Command{
+	Use:   "ethermintd",
```
Comment (jackzampolin) [inline cmd/ethermintd/root.go]: I think we are going to rename the binaries to `emintd` and `emintcli`

File: cmd/ethermintd/root.go [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+	Short: "Ethereum on Tendermint!",
+	Run: runRoot,
+}
+
+var config = &types.Config{}
+
+func init() {
+	rootCmd.Flags().BoolVarP(&config.EnableRPC, "rpc", "", false, "Enable the HTTP-RPC server")
```
Comment (jackzampolin) [inline cmd/ethermintd/root.go]: Can we enable loading these values from a config file as well?

File: server/apis.go [hunk 3]
```diff
@@ -0,0 +1,33 @@
+package server
+
+import (
+	"github.com/ethereum/go-ethereum/rpc"
```
Comment (jackzampolin) [inline server/apis.go]: `gofmt`

File: cmd/ethermintd/root.go [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+	rootCmd.Flags().StringVarP(&config.RPCAddr, "rpcaddr", "", "127.0.0.1", "HTTP-RPC server listening interface")
+	rootCmd.Flags().IntVarP(&config.RPCPort, "rpcport", "", 8545, "HTTP-RPC server listening port")
+	rootCmd.Flags().StringSliceVarP(&config.RPCCORSDomains, "rpccorsdomain", "", nil, "Comma-separated list of domains from which to accept cross origin requests")
+	rootCmd.Flags().StringSliceVarP(&config.RPCVHosts, "rpcvhosts", "", nil, "Comma-separated list of virtual hostnames from which to accept requests")
+}
+
+func runRoot(cmd *cobra.Command, args []string) {
```
Comment (jackzampolin) [inline cmd/ethermintd/root.go]: Like the server! This is quite nice with all the signal handling! 💅

File: server/start.go [hunk 5]
```diff
@@ -1 +1,32 @@
 package server
+
+import (
+	"github.com/ethereum/go-ethereum/rpc"
```
Comment (jackzampolin) [inline server/start.go]: `gofmt`

File: server/start_test.go [hunk 6]
```diff
@@ -0,0 +1,75 @@
+package server
+
+import (
+	"testing"
```
Comment (jackzampolin) [inline server/start_test.go]: `gofmt`

File: types/config.go [hunk 7]
```diff
@@ -0,0 +1,9 @@
+package types
+
+type Config struct {
```
Comment (jackzampolin) [inline types/config.go]: Add `json` (or preferably `yaml` [I think the `json` ones might work for that too...]) struct tags here so we can read the config in from a file.

Comment (jackzampolin) [reply types/config.go]: `~/.emintd.yaml` maybe?

File: version/version.go [hunk 8]
```diff
@@ -8,3 +13,8 @@ const Version = "0.0.0"
 
 // GitCommit contains the git SHA1 short hash set by build flags.
 var GitCommit = ""
+
+// ClientVersion returns the full version string for identification on the larger Ethereum network.
+func ClientVersion() string {
```
Comment (jackzampolin) [inline version/version.go:21]: Going to need to make sure this gets hooked up in the Makefile!

Comment (jackzampolin) [reply version/version.go:21]: We can keep that out of scope for this PR, and open an issue.

**Review by jackzampolin — COMMENTED** [review]
I have a couple of initial comments, but this looks like a great start. I think @alexanderbez might have some other comments here.

File: cmd/ethermintd/root.go (hunk 2 above)
Comment (mslipper) [reply cmd/ethermintd/root.go]: Yes, super easy to do with Viper and Cobra together - recommend doing that separately though so we can discuss what we want to name the config file, which filetypes to support, etc.

Comment (mslipper) [reply cmd/ethermintd/root.go]: Specifically, we can pull config from a variety of sources via Viper - CLI, environment variables, and config files of TOML/YAML/JSON/properties formats.

File: types/config.go (hunk 7 above)
Comment (mslipper) [reply types/config.go]: See above comment re: Viper. Happy to do implement this now as well, just let me know.

File: cmd/emintd/root.go [hunk 9]
```diff
@@ -0,0 +1,62 @@
+package main
```
Comment (alexanderbez) [inline cmd/emintd/root.go]: To be honest, I think we should avoid implementing any logic here until #436 is implemented as that will lay the groundwork for actually starting a TM backend, creating the app and starting things. Most of that will live in `server/` anyway.

Let's leave this file untouched for now (but feel free to keep it in your branch for testing purposes).

File: server/apis.go [hunk 10]
```diff
@@ -0,0 +1,33 @@
+package server
```
Comment (alexanderbez) [inline server/apis.go]: Missing proper godocs for all exported functions and fields 👍

File: server/apis_test.go [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+import (
+	"context"
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
+	"github.com/tendermint/ethermint/types"
+	"github.com/tendermint/ethermint/version"
+	"testing"
+)
+
+type APIsTestSuite struct {
```
Comment (alexanderbez) [inline server/apis_test.go]: I don't think `APIsTestSuite` needs to be public here, right?

File: server/start.go [hunk 12]
```diff
@@ -1 +1,32 @@
 package server
```
Comment (alexanderbez) [inline server/start.go]: Can we rename this to `rpc.go`? `start.go` will be for the actual app 👍

File: server/start.go [hunk 13]
```diff
@@ -1 +1,32 @@
...
+
+import (
+	"context"
+	"fmt"
+	"github.com/ethereum/go-ethereum/rpc"
+	"github.com/tendermint/ethermint/types"
+)
+
+func StartHTTPEndpoint(ctx context.Context, config *types.Config, apis []rpc.API) (*rpc.Server, error) {
```
Comment (alexanderbez) [inline server/start.go]: Missing godoc

File: server/start_test.go [hunk 14]
```diff
@@ -0,0 +1,75 @@
+package server
```
Comment (alexanderbez) [inline server/start_test.go]: `rpc_test.go`

File: server/start_test.go [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+	"github.com/stretchr/testify/require"
+	"github.com/tendermint/ethermint/types"
+	"io/ioutil"
+	"math/rand"
+	"net/http"
+	"strings"
+	"testing"
+)
+
+type TestService struct{}
```
Comment (alexanderbez) [inline server/start_test.go]: This doesn't need to be public does it?

File: server/start_test.go [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+
+func (s *TestService) Foo(arg string) string {
+	return arg
+}
+
+func TestStartHTTPEndpoint_StartStop(t *testing.T) {
```
Comment (alexanderbez) [inline server/start_test.go]: Avoid `_` 👍

File: version/version.go (hunk 8 above)
Comment (alexanderbez) [reply version/version.go:21]: Mhmmm, doesn't seem `ClientVersion` introduced any new build-based variables -- I think this is fine unless I'm missing something?

Comment (alexanderbez) [reply version/version.go:21]: On another note, maybe move `Ethermint` to a constant and and use `version.AppName` in `NewEthermintApp` as that also uses the same thing.

File: types/config.go (hunk 7 above)
Comment (alexanderbez) [reply types/config.go]: So Tendermint config will live under `<home>/config/config.toml`. Such a config for the app will live there too (e.g. `<home>/config/emint.toml`). I want to limit confusion for users as much as possible.

I think we need to define a convention for what our config will take, what it'll be named, and how/when it'll be used. For now though, we should move this file to `server/`.

**Review by alexanderbez — CHANGES_REQUESTED** [review]
Thanks for the contribution @mslipper! Great work. I left a few remarks -- ultimately around structure and what I believe this PR should ultimately introduce. Lmk your thoughts.

File: types/config.go (hunk 7 above)
Comment (jackzampolin) [reply types/config.go]: Sounds like this is the convention? Maybe we hold off on implementation here for another PR?

Comment (alexanderbez) [reply types/config.go]: ++

File: server/start_test.go (hunk 15 above)
Comment (mslipper) [reply server/start_test.go]: It does - the HTTP handler does reflection-based checks that the RPC struct is exported.

File: server/apis.go (hunk 10 above)
Comment (alexanderbez) [reply server/apis.go]: Sorry @mslipper don't wanna be nit-picky here, but can we follow Golang convention here (`<Name> description ...`).

File: server/apis.go [hunk 17]
```diff
@@ -0,0 +1,36 @@
...
+
+import (
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/rpc"
+	"github.com/tendermint/ethermint/version"
+)
+
+// returns the master list of public APIs for use with StartHTTPEndpoint
+func GetAPIs() []rpc.API {
```
Comment (alexanderbez) [inline server/apis.go]: Should this be `GetRPCAPIs`?

File: server/apis_test.go [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+	require.Nil(s.T(), err, "unexpected error")
+	s.Stop = stop
+	s.Port = port
+}
+
+func (s *apisTestSuite) TearDownSuite() {
+	s.Stop()
+}
+
+func (s *apisTestSuite) TestPublicWeb3API_ClientVersion() {
```
Comment (alexanderbez) [inline server/apis_test.go]: Should be no `_` in function/method names 👍

File: server/apis_test.go [hunk 19]
```diff
@@ -0,0 +1,55 @@
...
+	res, err := rpcCall(s.Port, "web3_clientVersion", []string{})
+	require.Nil(s.T(), err, "unexpected error")
+	require.Equal(s.T(), version.ClientVersion(), res)
+}
+
+func (s *apisTestSuite) TestPublicWeb3API_Sha3() {
```
Comment (alexanderbez) [inline server/apis_test.go]: ditto

File: server/config.go [hunk 20]
```diff
@@ -0,0 +1,14 @@
+package server
+
```
Comment (alexanderbez) [inline server/config.go]: ditto regarding godocs 👍

File: server/rpc.go [hunk 21]
```diff
@@ -0,0 +1,33 @@
...
+
+import (
+	"context"
+	"fmt"
+	"github.com/ethereum/go-ethereum/rpc"
+	)
+
+// starts the Tendermint Web3-compatible RPC layer. Consumes a Context for cancellation, a config
+// struct, and a list of rpc.API interfaces that will be automatically wired into a JSON-RPC webserver.
+func StartHTTPEndpoint(ctx context.Context, config *Config, apis []rpc.API) (*rpc.Server, error) {
```
Comment (alexanderbez) [inline server/rpc.go]: ditto regarding godocs 👍

**Review by alexanderbez — CHANGES_REQUESTED** [review]
@mslipper thanks for the updates! I'm starting to think RPC and related logic should live under `server/rpc` + also see other comments.

**Comment by alexanderbez** [comment]
@mslipper awesome -- looks like you addressed most of the feedback! I think there is still the notion of moving the rpc (and related logic) implementation to `server/rpc`...at least for now. We're going to have Tendermint RPC, and lite client REST logic there too I imagine.

**Comment by alexanderbez** [comment]
@mslipper once CI failure is addressed, I'll be happy to approve and merge 👍

**Comment by jackzampolin** [comment]
Looks like a linting issue.

**Review by alexanderbez — APPROVED** [review]
ACK -- thank you 👍

**Pull request merged**